### PR TITLE
Add postProcessRequest to SecurityService

### PR DESCRIPTION
**Pull request by cgtz**
This adds a new method to the SecurityService interface that is used to
run additional request security validations after the request has been
fully parsed (i.e. inbound id conversion has finished). This is a useful
place to implement AclService calls.

*Reviewers:* @xiahome @vgkholla

### CODE REVIEW

File: ambry-api/src/main/java/com.github.ambry/rest/SecurityService.java [hunk 1]
```diff
@@ -49,6 +59,18 @@
    * @return A future that would contain information about whether processing of request succeeded or not,
    * eventually.
    */
-  public Future<Void> processResponse(RestRequest restRequest, RestResponseChannel responseChannel, BlobInfo blobInfo,
+  Future<Void> processResponse(RestRequest restRequest, RestResponseChannel responseChannel, BlobInfo blobInfo,
```
Comment (vgkholla) [inline ambry-api/src/main/java/com.github.ambry/rest/SecurityService.java]: could we convert these functions too? If it isn't complicated.

File: ambry-frontend/src/main/java/com.github.ambry.frontend/AmbrySecurityService.java [hunk 2]
```diff
@@ -72,6 +72,20 @@ public AmbrySecurityService(FrontendConfig frontendConfig, FrontendMetrics front
...
+    Exception exception = null;
+    frontendMetrics.securityServicePostProcessRequestRate.mark();
+    long startTimeMs = System.currentTimeMillis();
+    if (!isOpen) {
+      exception = new RestServiceException("SecurityService is closed", RestServiceErrorCode.ServiceUnavailable);
+    } else if (restRequest == null || callback == null) {
+      throw new IllegalArgumentException("RestRequest or Callback is null");
+    }
+    callback.onCompletion(null, exception);
+    frontendMetrics.securityServicePostProcessRequestTimeInMs.update(System.currentTimeMillis() - startTimeMs);
```
Comment (vgkholla) [inline ambry-frontend/src/main/java/com.github.ambry.frontend/AmbrySecurityService.java]: this should be before the callback I think

File: ambry-utils/src/main/java/com.github.ambry.utils/Utils.java [hunk 3]
```diff
@@ -756,6 +756,18 @@ public static long getTimeInMsToTheNearestSec(long timeInMs) {
   }
 
   /**
+   * Extracts the cause of an {@link Exception}.
+   * @param e the {@link Exception}
+   * @return if the cause is {@code null}, return {@code e} itself. If the cause is not an instance
+   *         of exception, return the {@link Throwable} wrapped in an exception. Otherwise, return the cause
+   *         {@link Exception}.
+   */
+  public static Exception extractCause(Exception e) {
```
Comment (vgkholla) [inline ambry-utils/src/main/java/com.github.ambry.utils/Utils.java]: can me move this from utils as discussed?

File: ambry-frontend/src/test/java/com.github.ambry.frontend/AmbryBlobStorageServiceTest.java [hunk 4]
```diff
@@ -1424,21 +1424,15 @@ public void close() {
     }
 
     /**
-     * Completes the operation by creating and invoking a {@link Future} and invoking the {@code callback} if non-null.
-     * @param callback the {@link Callback} to invoke. Can be null.
+     * Completes the operation by  invoking the {@code callback}.
```
Comment (vgkholla) [inline ambry-frontend/src/test/java/com.github.ambry.frontend/AmbryBlobStorageServiceTest.java]: minor: extra space b/w by and invoking

File: ambry-frontend/src/test/java/com.github.ambry.frontend/AmbrySecurityServiceTest.java [hunk 5]
```diff
@@ -487,19 +487,11 @@ private void testExceptionCasesProcessResponse(RestMethod restMethod, RestRespon
...
-      Assert.assertNotNull("Exception should have been thrown", callback.exception);
-      re = (RestServiceException) callback.exception;
-      Assert.assertEquals("Unexpected RestServerErrorCode (Callback)", expectedErrorCode, re.getErrorCode());
-    }
+    securityService.processResponse(restRequest, restResponseChannel, blobInfo, callback);
+    Assert.assertTrue("Callback should have been invoked", callback.callbackLatch.await(1, TimeUnit.SECONDS));
+    Assert.assertNotNull("Exception should have been thrown", callback.exception);
+    RestServiceException re = (RestServiceException) callback.exception;
+    Assert.assertEquals("Unexpected RestServerErrorCode (Callback)", expectedErrorCode, re.getErrorCode());
   }
```
Comment (vgkholla) [inline ambry-frontend/src/test/java/com.github.ambry.frontend/AmbrySecurityServiceTest.java]: if it isn't too much change, lets get rid of `SecurityServiceCallback`

File: ambry-frontend/src/main/java/com.github.ambry.frontend/AmbrySecurityService.java [hunk 6]
```diff
@@ -62,17 +61,26 @@ public AmbrySecurityService(FrontendConfig frontendConfig, FrontendMetrics front
...
     frontendMetrics.securityServiceProcessRequestTimeInMs.update(System.currentTimeMillis() - startTimeMs);
-    return futureResult;
+    callback.onCompletion(null, exception);
   }
 
   @Override
-  public Future<Void> processResponse(RestRequest restRequest, RestResponseChannel responseChannel, BlobInfo blobInfo,
+  public void postProcessRequest(RestRequest restRequest, Callback<Void> callback) {
+    Exception exception = null;
+    frontendMetrics.securityServicePostProcessRequestRate.mark();
```
Comment (xiahome) [inline ambry-frontend/src/main/java/com.github.ambry.frontend/AmbrySecurityService.java]: May I know the reason you want to track this? Would this be `totalRequestRate - totalPutRate`?

File: ambry-utils/src/test/java/com.github.ambry.utils/TestUtils.java [hunk 7]
```diff
@@ -14,9 +14,12 @@
 package com.github.ambry.utils;
 
 import java.util.Random;
+import java.util.concurrent.Callable;
```
Comment (xiahome) [inline ambry-utils/src/test/java/com.github.ambry.utils/TestUtils.java]: Seems not needed.

File: ambry-utils/src/test/java/com.github.ambry.utils/TestUtils.java [hunk 8]
```diff
@@ -81,4 +84,53 @@ public static void awaitLatchOrTimeout(CountDownLatch latch, long timeoutMs)
       throw new TimeoutException("Too long time to complete operation.");
     }
   }
+
+  /**
+   * Succeed if the {@code body} throws an exception of type {@code exceptionClass}, otherwise fail.
+   * @param exceptionClass the type of exception that should occur.
+   * @param body the body to execute. This should throw an exception of type {@code exceptionClass}
+   * @param errorAction if non-null and the exception class matches, execute this action.
+   */
```
Comment (xiahome) [inline ambry-utils/src/test/java/com.github.ambry.utils/TestUtils.java]: add `@throws`

File: ambry-frontend/src/main/java/com.github.ambry.frontend/GetPeersHandler.java [hunk 9]
```diff
@@ -100,6 +149,52 @@ void handle(RestRequest restRequest, RestResponseChannel restResponseChannel,
...
     /**
+     * If {@code exception} is null, call {@link SecurityService#postProcessRequest(RestRequest, Callback)}.
+     * @param result The result of the request. This would be non null when the request executed successfully
+     * @param exception The exception that was reported on execution of the request
+     */
+    @Override
+    public void onCompletion(Void result, Exception exception) {
+      long processingStartTimeMs = SystemTime.getInstance().milliseconds();
+      metrics.getPeersSecurityRequestTimeInMs.update(processingStartTimeMs - operationStartTimeMs);
+      ReadableStreamChannel channel = null;
```
Comment (xiahome) [inline ambry-frontend/src/main/java/com.github.ambry.frontend/GetPeersHandler.java]: Seems not being used.

File: ambry-frontend/src/main/java/com.github.ambry.frontend/AmbrySecurityService.java (hunk 6 above)
Comment (cgtz) [inline ambry-frontend/src/main/java/com.github.ambry.frontend/AmbrySecurityService.java]: Mostly adding this b/c similar metrics are in place for `processRequest` and `processResponse`. This should be equal to the number of total requests that pass processRequest and idMapping.

File: ambry-frontend/src/main/java/com.github.ambry.frontend/GetPeersHandler.java (hunk 9 above)
Comment (cgtz) [inline ambry-frontend/src/main/java/com.github.ambry.frontend/GetPeersHandler.java]: removed

File: ambry-utils/src/test/java/com.github.ambry.utils/TestUtils.java (hunk 7 above)
Comment (cgtz) [inline ambry-utils/src/test/java/com.github.ambry.utils/TestUtils.java]: removed

File: ambry-utils/src/test/java/com.github.ambry.utils/TestUtils.java (hunk 8 above)
Comment (cgtz) [inline ambry-utils/src/test/java/com.github.ambry.utils/TestUtils.java]: added

**Comment by xiahome** [comment]
LGTM.

**Pull request merged**